book/testing_code.md

```diff
@@ -321,12 +321,65 @@ Integration tests that check very specific outcomes will need to be updated with
-Test that your code works, given the format of response that the system can give.
-Mocks?
+## Isolate code tests from external systems
+
+Testing code that interacts with an external system can be particularly challenging when you can't guarantee
+that the system will provide you with the same response each time; this could include code querying a database or
+making API requests, for example.
+
+Best practice is to seperate external system dependencies from your code tests as much as possible. This can mitigate against various risks, depending on your application:
+
+- Testing database interaction with a production database could result in damage to, or loss of, data.
+- Making real API calls when testing a function that handles requests could incur unintended monetary costs.
+
+Isolating code from external systems allows for tests to run without reliance on the real systems; for example, tests for a database interaction that can still run even if the database connection goes down.
+Testing your code in this way means that tests evaluate how your code handles an output or response, and not the system dependency itself. This has the added benefit of helping you understand when errors are coming from an external system, as opposed to your code.
+
+One way of achieving this is with mocking, where a response from an outside system is replaced with a mock object that your code can be tested against. In this example, there's a function making an API request in `src/handle_api_request.py`, and two test functions in `tests/test_handle_api_request.py`. The response from `requests.get()` is mocked with a `Mock()` object, to which `text` and `status_code` attributes are assigned, so that the `get_response()` function can be evaluated for how it handles successful and unsuccessful requests. Thanks to the mocking, get requests are not made to `http://example.com`. AI has been used to produce content within this artefact.
+
+```{code-block} python
+# src/handle_api_request.py
+import requests
+
+def get_response(url: str):
+    response = requests.get(url)
+    if response.status_code != 200:
+        raise(requests.HTTPError("Unsuccessful request"))
+
+    return response.text
+
+...
+
+# tests/test_handle_api_request.py
+from src.api_requests import get_response
+import requests
+import pytest
+from unittest import mock
+
+@mock.patch("requests.get")
+def test_get_response_success(mock_requests_get):
+    mock_response = mock.Mock()
+    mock_response.text = "Successful"
+    mock_response.status_code = 200
+
+    mock_requests_get.return_value = mock_response
+    
+    actual = get_response("http://example.com")
+    assert(actual == "Successful")
+
+@mock.patch("requests.get")
+def test_get_response_fail(mock_requests_get):
+    mock_response = mock.Mock()
+    mock_response.status_code = 400
+    mock_requests_get.return_value = mock_response
+    
+    with pytest.raises(requests.HTTPError):
+        actual = get_response("http://example.com")
+
```

Comment:
The assertions in this test would be a little too loose. If the mock did not happen at all, this test would pass and a request to example.com would fire. Interestingly, this domain exists too (I was not aware of that...) This can be demonstrated with the below test, which passes:

```
def test_get_response_fail_no_mock():    
    with pytest.raises(requests.HTTPError):
        get_response("http://example.com/doesnotexist")
```

A minor tweak to increase the stringency of the test is to match the expected pattern that is raised in our source code, like so:

```
@mock.patch("requests.get")
def test_get_response_fail(mock_requests_get):
    mock_response = mock.Mock()
    mock_response.status_code = 400
    mock_requests_get.return_value = mock_response
    
    with pytest.raises(requests.HTTPError, match="Unsuccessful request"):
        get_response("http://example.com")
```

That way, even if `requests.get()` raised a request, it would not match the expected exception message and still fail.

Reply:
Adding in verification statements would then increase the stringency of the test:

```
@mock.patch("requests.get")
def test_get_response_fail(mock_requests_get):
    mock_response = mock.Mock()
    mock_response.status_code = 400
    mock_requests_get.return_value = mock_response
    
    with pytest.raises(requests.HTTPError, match="Unsuccessful request"):
        get_response("http://example.com")
    mock_requests_get.assert_called()
    mock_requests_get.assert_called_once_with("http://example.com")
```

Reply:
This is not as simple to explain as your original code. It may be worth introducing these validation assertions in an additional iteration of the test code? EG - introduce the original code as is (though with the match parameter) then add the validation checks in an additional step with an explanation about their utility?